cpp/include/cudf/quantiles.hpp

```diff
@@ -63,21 +43,25 @@ quantile(column_view const& input,
  * @param input           Table used to compute quantile rows.
  * @param q               Desired quantiles in range [0, 1].
  * @param interp          Strategy used to select between the two rows on either
-                          side of the desired quantile.
- * @param sorted          Indicates if the input has been pre-sorted.
- * @param column_order    The desired sort order for each column.
- * @param null_precedence The desired order of null compared to other elements.
+ *                        side of the desired quantile.
+ * @param ordered_indices a column of integer indices that order `input`. If
+                          empty, `input` is assumed to be sorted.
+ * @param retain_types    Indicates if the result should be cast to the input
```

Comment:
There is at least the case of `int64_t`, where underflow and precision loss occur if casting to double and then back again. I will add an explicit test case for this scenario.

https://github.com/rapidsai/cudf/issues/1417

Reply:
Can you explain a little more about what the problem is and how `retain_types` alleviates it?

Reply:
I'm adding a test case which should make it clear, but this may provide some context:

https://github.com/rapidsai/cudf/blob/branch-0.13/cpp/src/quantiles/quantiles_util.hpp#L81

Reply:
Oh I think the part that is not obvious is that `intepolate::midpoint` has a template specialization which can handle `int64_t` underflow.

Reply:
It's also possible to add a specialization for `interpolate::linear` which can enhance precision when for `int64_t`, but this PR does not include those changes.

Reply:
Okay, but this still looks more complicated then it needs to be.

Why is there a templated overload for `int64_t` inputs _and_ a specialization for `int64_t` inputs/return. 

This looks a lot simpler to me and has the added benefit of not needing to explicit `Result` template parameter: https://wandbox.org/permlink/XJ6yIwOCUuWR0E32

Reply:
Which is good if you want the return type to be `int64_t`, but if you want a `double`, you'll lose precision by returning `int64_t` and then casting to `double`.

https://godbolt.org/z/2Ss-Jz

Reply:
https://godbolt.org/z/u4iGKF

Reply:
I consider the template argument here as an instruction to the algorithm to "do your best to give me the right result", which is something the algorithm needs to know explicitly, it seems. :man_shrugging: